File: avicola-backend/src/main/java/com/avicola/service/GalponService.java
package com.avicola.service;

import com.avicola.model.Galpon;
import com.avicola.repository.GalponRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.List;

// ─────────────────────────────────────────────────────────────────
//  SERVICIO: GalponService
//
//  Contiene toda la lógica de negocio de los galpones.
//  El controller llama a este servicio; este servicio llama
//  al repositorio para hablar con la base de datos.
//
//  Flujo: Controller → Service → Repository → Base de Datos
// ─────────────────────────────────────────────────────────────────
@Service
@RequiredArgsConstructor
public class GalponService {

    private final GalponRepository galponRepository;

    // Devuelve todos los galpones
    public List<Galpon> listarTodos() {
        return galponRepository.findAll();
    }

    // Busca un galpón por su ID. Si no existe, lanza error.
    public Galpon buscarPorId(Long id) {
        return galponRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Galpón no encontrado con ID: " + id));
    }

    // Crea un galpón nuevo. Verifica que el número no esté repetido.
    public Galpon crear(Galpon galpon) {
        if (galponRepository.existsByNumero(galpon.getNumero())) {
            throw new RuntimeException("Ya existe un galpón con el número: " + galpon.getNumero());
        }
        return galponRepository.save(galpon);
    }

    // Actualiza los datos de un galpón existente
    public Galpon actualizar(Long id, Galpon datos) {
        Galpon galpon = buscarPorId(id);
        galpon.setNumero(datos.getNumero());
        galpon.setNumeroAves(datos.getNumeroAves());
        galpon.setRaza(datos.getRaza());
        galpon.setEstado(datos.getEstado());
        return galponRepository.save(galpon);
    }

    // Elimina un galpón por su ID
    public void eliminar(Long id) {
        buscarPorId(id); // Verifica que existe antes de eliminar
        galponRepository.deleteById(id);
    }

    // Descuenta aves del galpón cuando se registra mortalidad
    public void reducirAves(Long id, int cantidad) {
        Galpon galpon = buscarPorId(id);
        int nuevo = Math.max(0, galpon.getNumeroAves() - cantidad); // No puede ser negativo
        galpon.setNumeroAves(nuevo);
        galponRepository.save(galpon);
    }
}

File: avicola-frontend/src/services/galponService.js
import { api } from './api.js'

/**
 * Servicio para gestionar galpones
 * Tabla: galpon (id_galpon, numero_Galpon, numero_aves)
 */
export const galponService = {
  // Obtener todos los galpones
  async getAll() {
    try {
      const response = await api.get('/galpones')
      return response.galpones || response
    } catch (error) {
      console.error('Error obteniendo galpones:', error)
      throw error
    }
  },

  // Obtener un galpon por ID
  async getById(id) {
    try {
      const response = await api.get(`/galpones/${id}`)
      return response.galpon || response
    } catch (error) {
      console.error(`Error obteniendo galpón ${id}:`, error)
      throw error
    }
  },

  // Crear nuevo galpón
  async create(galponData) {
    try {
      const response = await api.post('/galpones', galponData)
      return response.galpon || response
    } catch (error) {
      console.error('Error creando galpón:', error)
      throw error
    }
  },

  // Actualizar galpón existente
  async update(id, galponData) {
    try {
      const response = await api.put(`/galpones/${id}`, galponData)
      return response.galpon || response
    } catch (error) {
      console.error(`Error actualizando galpón ${id}:`, error)
      throw error
    }
  },

  // Eliminar galpón
  async delete(id) {
    try {
      await api.delete(`/galpones/${id}`)
      return true
    } catch (error) {
      console.error(`Error eliminando galpón ${id}:`, error)
      throw error
    }
  },

  // Obtener galpones con capacidad disponible
  async getWithCapacity() {
    try {
      const response = await api.get('/galpones/disponibles')
      return response.galpones || response
    } catch (error) {
      console.error('Error obteniendo galpones disponibles:', error)
      throw error
    }
  },

  // Actualizar número de aves en galpón
  async updateBirdCount(id, numeroAves) {
    try {
      const response = await api.patch(`/galpones/${id}/aves`, { numero_aves: numeroAves })
      return response.galpon || response
    } catch (error) {
      console.error(`Error actualizando conteo de aves en galpón ${id}:`, error)
      throw error
    }
  }
}
